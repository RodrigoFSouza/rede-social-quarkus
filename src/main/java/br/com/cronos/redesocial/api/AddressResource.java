package br.com.cronos.redesocial.api;

import br.com.cronos.redesocial.api.dto.CreateAddressRequest;
import br.com.cronos.redesocial.domain.model.Address;
import br.com.cronos.redesocial.domain.model.User;
import io.quarkus.panache.common.Sort;

import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.Optional;

@Path("/api/v1/users/")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class AddressResource {
    @GET
    @Path("{userId}/address")
    public Response listAllAdrress(@PathParam("userId") Long userId) {
        Optional<User> userOp = User.findByIdOptional(userId);
        if (userOp.isEmpty()) {
            throw new NotFoundException("User not found");
        }

        List<Address> results = Address.find("user", Sort.by("id"), userOp.get()).list();
        return Response.ok(results).build();
    }

    @GET
    @Path("/address")
    public Response listAllAdrress() {
        List<Address> results = Address.listAll();
        return Response.ok(results).build();
    }

    @POST
    @Path("{userId}/address")
    @Transactional
    public Response createAddress(@PathParam("userId") Long userId, CreateAddressRequest request) {
        Optional<User> userOp = User.findByIdOptional(userId);
        if (userOp.isEmpty()) {
            throw new NotFoundException("User not found");
        }
        var address = new Address();
        address.setStreet(request.getStreet());
        address.setNumber(request.getNumber());
        address.setCity(request.getCity());
        address.setDistrict(request.getDistrict());
        address.setState(request.getState());
        address.setZipcode(request.getZipcode());
        address.setUser(userOp.get());
        Address.persist(address);

        return Response.status(Response.Status.CREATED).build();
    }

    @PUT
    @Path("{userId}/address/{id}")
    @Transactional
    public Response updateAddress(@PathParam("userId") Long userId, @PathParam("id") Long id, Address address) {
        Optional<Address> addressOp = Address.findByIdOptional(id);
        if (addressOp.isEmpty()) {
            throw new NotFoundException();
        }

        Optional<User> userOp = User.findByIdOptional(userId);
        if (userOp.isEmpty()) {
            throw new NotFoundException("User not found");
        }

        var addressUpdate = addressOp.get();
        if (!userOp.get().getId().equals(addressUpdate.getId())) {
            throw new BadRequestException("User this reques is diferent of address persisted");
        }

        addressUpdate.setCity(address.getCity());
        addressUpdate.setState(address.getState());
        addressUpdate.setStreet(address.getStreet());
        addressUpdate.setZipcode(address.getZipcode());
        addressUpdate.setNumber(address.getNumber());
        addressUpdate.setDistrict(address.getDistrict());

        Address.persist(addressUpdate);

        return Response.noContent().build();
    }

    @DELETE
    @Path("{userId}/address/{id}")
    @Transactional
    public Response deleteAddress(@PathParam("userId") Long userId, @PathParam("id") Long id) {
        Optional<User> userOp = User.findByIdOptional(id);
        if (userOp.isEmpty()) {
            throw new NotFoundException("User not found");
        }

        Optional<Address> addressOp = Address.findByIdOptional(id);
        addressOp.ifPresentOrElse(Address::delete, () -> { throw new NotFoundException(); });

        return Response.noContent().build();
    }
}
