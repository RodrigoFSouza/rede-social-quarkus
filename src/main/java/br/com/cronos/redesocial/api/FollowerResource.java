package br.com.cronos.redesocial.api;

import br.com.cronos.redesocial.api.dto.FollowerRequest;
import br.com.cronos.redesocial.domain.model.Follower;
import br.com.cronos.redesocial.domain.repository.FollowerRepository;
import br.com.cronos.redesocial.domain.repository.UserRepository;

import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Path("/api/v1/users/{userId}/followers")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class FollowerResource {

    private final FollowerRepository followerRepository;
    private final UserRepository userRepository;

    @Inject
    public FollowerResource(FollowerRepository followerRepository, UserRepository userRepository) {
        this.followerRepository = followerRepository;
        this.userRepository = userRepository;
    }

    @PUT
    @Transactional
    public Response follower(@PathParam("userId") Long userId, FollowerRequest request) {
        if (userId.equals(request.getFollowerId())) {
            return Response.status(Response.Status.CONFLICT)
                    .entity("You can't follower yorself")
                    .build();
        }

        var user = userRepository.findById(userId);

        if (user == null) {
            return Response
                    .status(Response.Status.NOT_FOUND)
                    .entity("User not found of id " + userId)
                    .build();
        }

        var follower = userRepository.findByIdOptional(request.getFollowerId());
        if (follower.isEmpty()) {
            return Response
                    .status(Response.Status.NOT_FOUND)
                    .entity("Follower not found of id " + request.getFollowerId())
                    .build();
        }

        boolean follows = followerRepository.follows(follower.get(), user);

        if (!follows) {
            var entity = new Follower();
            entity.setUser(user);
            entity.setFollower(follower.get());
            followerRepository.persist(entity);
        }

        return Response.status(Response.Status.NO_CONTENT).build();
    }
}
