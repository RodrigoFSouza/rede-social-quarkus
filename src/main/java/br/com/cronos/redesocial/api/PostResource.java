package br.com.cronos.redesocial.api;

import br.com.cronos.redesocial.api.dto.CreatePostRequest;
import br.com.cronos.redesocial.api.dto.PostResponse;
import br.com.cronos.redesocial.domain.model.Post;
import br.com.cronos.redesocial.domain.model.User;
import br.com.cronos.redesocial.domain.repository.FollowerRepository;
import br.com.cronos.redesocial.domain.repository.PostRepository;
import br.com.cronos.redesocial.domain.repository.UserRepository;
import io.quarkus.panache.common.Sort;

import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.stream.Collectors;

import static java.util.Objects.isNull;

@Path("/api/v1/users/{userId}/posts")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class PostResource {

    private final UserRepository userRepository;
    private final PostRepository postRepository;
    private final FollowerRepository followerRepository;

    public PostResource(UserRepository userRepository, PostRepository postRepository, FollowerRepository followerRepository) {
        this.userRepository = userRepository;
        this.postRepository = postRepository;
        this.followerRepository = followerRepository;
    }

    @POST
    @Transactional
    public Response savePost(@PathParam("userId") Long userId, CreatePostRequest createPostRequest) {
        User user = userRepository.findById(userId);

        if (user == null) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }
        Post post = new Post();
        post.setText(createPostRequest.getText());
        post.setUser(user);

        postRepository.persist(post);
        return Response.status(Response.Status.CREATED).build();
    }

    @GET
    public Response listPosts(@PathParam("userId") Long userId, @HeaderParam("followerId") Long followerId) {
        User user = userRepository.findById(userId);

        if (user == null) {
            return Response.status(Response.Status.NOT_FOUND).build();
        }

        if (isNull(followerId)) {
            return Response.status(Response.Status.BAD_REQUEST)
                    .entity("You forgot the header followerId")
                    .build();
        }

        User follower = userRepository.findById(followerId);

        if (isNull(follower)) {
            return Response.status(Response.Status.BAD_REQUEST)
                    .entity("Inexistent followerId")
                    .build();
        }

        boolean follows = followerRepository.follows(follower, user);
        if (!follows) {
            return Response.status(Response.Status.FORBIDDEN)
                    .entity("You canÂ´t see these posts")
                    .build();
        }

        var query = postRepository.find("user", Sort.by("dateTime", Sort.Direction.Descending) ,user);
        var list = query.list();

        List<PostResponse> posts = list
                .stream()
                .map(PostResponse::fromDto)
                .collect(Collectors.toList());
        return Response.ok(posts).build();
    }
}
