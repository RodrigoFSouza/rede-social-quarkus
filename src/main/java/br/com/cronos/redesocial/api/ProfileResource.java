package br.com.cronos.redesocial.api;

import br.com.cronos.redesocial.api.dto.CreateProfileRequest;
import br.com.cronos.redesocial.api.dto.CreateUserRequest;
import br.com.cronos.redesocial.domain.model.Profile;
import br.com.cronos.redesocial.domain.repository.ProfileRepository;
import io.quarkus.hibernate.orm.panache.PanacheQuery;

import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Path("/api/v1/profiles")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class ProfileResource {

    private final ProfileRepository profileRepository;

    @Inject
    public ProfileResource(ProfileRepository profileRepository) {
        this.profileRepository = profileRepository;
    }

    @POST
    @Transactional
    public Response createProfile(CreateProfileRequest profileRequest) {
        Profile profile = new Profile();
        profile.setName(profileRequest.getName());

        profileRepository.persist(profile);

        return Response.ok(profile).build();
    }

    @GET
    public Response listAllProfiles() {
        PanacheQuery<Profile> query = profileRepository.findAll();
        return Response.ok(query.list()).build();
    }

    @DELETE
    @Path("{id}")
    @Transactional
    public Response deleteProfile(@PathParam("id") Long id) {
        Profile profile = profileRepository.findById(id);

        if (profile != null) {
            profileRepository.delete(profile);
            return Response.ok().build();
        }
        return Response.status(Response.Status.NOT_FOUND).build();
    }

    @PUT
    @Path("{id}")
    @Transactional
    public Response updateUser(@PathParam("id") Long id, CreateUserRequest userData) {
        Profile profile = profileRepository.findById(id);

        if (profile != null) {
            profile.setName(userData.getName());

            return Response.ok().build();
        }
        return Response.status(Response.Status.NOT_FOUND).build();
    }
}
