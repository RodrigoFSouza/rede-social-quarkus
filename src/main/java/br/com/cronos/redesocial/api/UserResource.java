package br.com.cronos.redesocial.api;

import br.com.cronos.redesocial.api.dto.CreateUserRequest;

import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.Arrays;
import java.util.List;

@Path("/api/v1/users")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class UserResource {

    @POST
    public Response createUser( CreateUserRequest userRequest) {
        return Response.ok(userRequest).build();
    }

    @GET
    public Response listAllUsers() {
      return Response.ok(Arrays.asList(new CreateUserRequest())).build();
    }
}
