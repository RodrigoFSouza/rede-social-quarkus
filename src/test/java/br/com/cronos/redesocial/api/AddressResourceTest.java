package br.com.cronos.redesocial.api;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.dataset.DataSet;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.ContentType;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import static io.restassured.RestAssured.given;

@DBRider
@QuarkusTest
@QuarkusTestResource(RedeSocialTestLifecycleManager.class)
@DBUnit(schema = "public", caseSensitiveTableNames = false, cacheConnection = false)
public class AddressResourceTest {
    @Test
    @DataSet(value = "datasets/address.yml")
    public void listAllAddressFromUserTest() {
        var response = given()
            .contentType(ContentType.JSON)
        .when()
            .get("/api/v1/users/address")
        .then()
                .extract().response();

        Approvals.verify(response.asString());
    }
}
