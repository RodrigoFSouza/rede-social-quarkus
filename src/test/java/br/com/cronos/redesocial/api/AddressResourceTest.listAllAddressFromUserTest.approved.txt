[{"city":"Maringá","district":"Centro","id":1,"number":"99","state":"Paraná","street":"Av. Brasil","zipcode":"87083-090"},{"city":"Maringá","district":"Centro","id":2,"number":"101","state":"Paraná","street":"Av. Paraná","zipcode":"87083-100"},{"city":"Curitiba","district":"Boqueirão","id":3,"number":"102","state":"Paraná","street":"Av. Colombo","zipcode":"87083-120"}]